Tekton on OpenShift: authentication issue with internal OpenShift registry

From https://github.com/GoogleContainerTools/jib/issues/2074#issuecomment-579921534

> I did a new test using the following parameters top of ocp3 and that fails using jib-maven 2.0.0
>
> ```
>     - name: build-and-push
>       image: gcr.io/cloud-builders/mvn
>       command:
>         - mvn
>         - compile
>         - com.google.cloud.tools:jib-maven-plugin:2.0.0:build
>         - -Djib.from.image=registry.redhat.io/redhat-openjdk-18/openjdk18-openshift
>         - -Djib.allowInsecureRegistries=true
>         - -Djib.from.auth.username=my-user
>         - -Djib.from.auth.password=my-pwd
>         - -Dimage=$(outputs.resources.image.url)
> ```

> If I add to the config too `-DsendCredentialsOverHttp=true`, then we got this error
> ```
> [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.0.0:build (default-cli) on project rest-http: 
> Build image failed: Failed to authenticate with registry docker-registry.default.svc:5000/test/sb-image 
> because: server did not return 'WWW-Authenticate: Bearer' header: Basic 
> realm=openshift,error="access denied" -> [Help 1]
> ```

The issue is migrated from #2074. Some highlights:

---
* From https://github.com/GoogleContainerTools/jib/issues/2074#issuecomment-579931427

> ```
> [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.0.0:build (default-cli) on project rest-http: 
> Build image failed: Failed to authenticate with registry docker-registry.default.svc:5000/test/sb-image 
> because: server did not return 'WWW-Authenticate: Bearer' header: Basic 
> realm=openshift,error="access denied" -> [Help 1]
> ```
>
> This is weird. This can happen in the following case:
>
> 1. The registry (`docker-registry.default.svc:5000`) initially requested and accepted bearer authentication. (That is, the registry returned `WWW-Authenticate: Bearer ...`). The registry successfully returned a working auth token.
> 2. Jib uses the returned token for subsequent registry API calls. The token works for a while.
> 3. Sometime later, the registry returned "401 Unauthorized" when Jib tried to use the same token, most likely because the token expired after 5 minutes. (I can see the Maven build is taking more than > 5 minutes, although I don't know the exact expiration time of the token.)
> 4. Jib now tries to refresh the expired bearer token, but the server returned `WWW-Authenticate: Basic ...` instead of `WWW-Authenticate: Brearer ...`.
>
> This feels like the server is not returning the right `WWW-Authenticate` header. The server is supposed to return `Bearer` instead of `Basic` in this case.

---
* From https://github.com/GoogleContainerTools/jib/issues/2074#issuecomment-579980276

> The registry is an [Integrated OpenShift V3 registry](https://hub.docker.com/r/openshift/origin-docker-registry).
> ```
>       image: 'openshift/origin-docker-registry:v3.11.0'
> ```

---
* From https://github.com/GoogleContainerTools/jib/issues/2074#issuecomment-579978683

> We also get this error even if the maven builds is about `15s`

--- a/jib-core/CHANGELOG.md
+++ b/jib-core/CHANGELOG.md
@@ -9,6 +9,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed authentication failure with error `server did not return 'WWW-Authenticate: Bearer' header` in certain cases (for example, on OpenShift). ([#2258](https://github.com/GoogleContainerTools/jib/issues/2258))
+
 ## 0.13.0
 
 ### Added

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryClient.java
@@ -262,6 +262,8 @@ public class RegistryClient {
   // mutable
   private final AtomicReference<Authorization> authorization = new AtomicReference<>();
   private boolean readOnlyBearerAuth;
+  private final AtomicReference<RegistryAuthenticator> initialBearerAuthenticator =
+      new AtomicReference<>();
 
   /**
    * Instantiate with {@link #factory}.
@@ -344,6 +346,7 @@ public class RegistryClient {
       return false; // server returned "WWW-Authenticate: Basic ..."
     }
 
+    initialBearerAuthenticator.set(authenticator.get());
     if (readOnlyBearerAuth) {
       authorization.set(authenticator.get().authenticatePull(credential));
     } else {
@@ -375,10 +378,13 @@ public class RegistryClient {
       }
     }
 
-    throw new RegistryAuthenticationFailedException(
-        registry,
-        repository,
-        "server did not return 'WWW-Authenticate: Bearer' header: " + wwwAuthenticate);
+    eventHandlers.dispatch(
+        LogEvent.debug(
+            "server did not return 'WWW-Authenticate: Bearer' header. Actual: " + wwwAuthenticate));
+    if (readOnlyBearerAuth) {
+      return Verify.verifyNotNull(initialBearerAuthenticator.get()).authenticatePull(credential);
+    }
+    return Verify.verifyNotNull(initialBearerAuthenticator.get()).authenticatePush(credential);
   }
 
   /**

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryClientTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryClientTest.java
@@ -20,6 +20,7 @@ import com.google.cloud.tools.jib.api.Credential;
 import com.google.cloud.tools.jib.api.DescriptorDigest;
 import com.google.cloud.tools.jib.api.LogEvent;
 import com.google.cloud.tools.jib.api.RegistryException;
+import com.google.cloud.tools.jib.api.RegistryUnauthorizedException;
 import com.google.cloud.tools.jib.blob.BlobDescriptor;
 import com.google.cloud.tools.jib.event.EventHandlers;
 import com.google.cloud.tools.jib.global.JibSystemProperties;
@@ -166,6 +167,71 @@ public class RegistryClientTest {
     Mockito.verify(eventHandlers).dispatch(logContains("bearer auth succeeded"));
     Mockito.verify(eventHandlers, Mockito.times(2))
         .dispatch(logContains("refreshing bearer auth token"));
+  }
+
+  @Test
+  public void testAutomaticTokenRefresh_badWwwAuthenticateResponse()
+      throws IOException, InterruptedException, GeneralSecurityException, URISyntaxException,
+          RegistryException {
+    String tokenResponse = "HTTP/1.1 200 OK\nContent-Length: 26\n\n{\"token\":\"awesome-token!\"}";
+    authServer = new TestWebServer(false, Arrays.asList(tokenResponse), 3);
+
+    List<String> responses =
+        Arrays.asList(
+            "HTTP/1.1 401 Unauthorized\nContent-Length: 0\nWWW-Authenticate: Bearer realm=\""
+                + authServer.getEndpoint()
+                + "\"\n\n",
+            "HTTP/1.1 401 Unauthorized\nContent-Length: 0\nWWW-Authenticate: Basic realm=foo\n\n",
+            "HTTP/1.1 401 Unauthorized\nContent-Length: 0\n\n",
+            "HTTP/1.1 200 OK\nContent-Length: 5678\n\n");
+    registry = new TestWebServer(false, responses, responses.size(), true);
+
+    RegistryClient registryClient = createRegistryClient(null);
+    Assert.assertTrue(registryClient.doPushBearerAuth());
+
+    Optional<BlobDescriptor> digestAndSize = registryClient.checkBlob(digest);
+    Assert.assertEquals(5678, digestAndSize.get().getSize());
+
+    // Verify authServer returned bearer token three times (i.e., refreshed twice)
+    Assert.assertEquals(3, authServer.getTotalResponsesServed());
+    Assert.assertEquals(4, registry.getTotalResponsesServed());
+
+    Mockito.verify(eventHandlers)
+        .dispatch(
+            logContains("server did not return 'WWW-Authenticate: Bearer' header. Actual: Basic"));
+    Mockito.verify(eventHandlers)
+        .dispatch(
+            logContains("server did not return 'WWW-Authenticate: Bearer' header. Actual: null"));
+  }
+
+  @Test
+  public void testAutomaticTokenRefresh_refreshLimit()
+      throws IOException, InterruptedException, GeneralSecurityException, URISyntaxException,
+          RegistryException {
+    String tokenResponse = "HTTP/1.1 200 OK\nContent-Length: 26\n\n{\"token\":\"awesome-token!\"}";
+    authServer = new TestWebServer(false, Arrays.asList(tokenResponse), 5);
+
+    String bearerAuth =
+        "HTTP/1.1 401 Unauthorized\nContent-Length: 0\nWWW-Authenticate: Bearer realm=\""
+            + authServer.getEndpoint()
+            + "\"\n\n";
+    String unauthorized = "HTTP/1.1 401 Unauthorized\nContent-Length: 0\n\n";
+    List<String> responses =
+        Arrays.asList(
+            bearerAuth, unauthorized, unauthorized, unauthorized, unauthorized, unauthorized);
+    registry = new TestWebServer(false, responses, responses.size(), true);
+
+    RegistryClient registryClient = createRegistryClient(null);
+    Assert.assertTrue(registryClient.doPushBearerAuth());
+
+    try {
+      registryClient.checkBlob(digest);
+      Assert.fail("Should have given up refreshing after 4 attempts");
+    } catch (RegistryUnauthorizedException ex) {
+      Assert.assertEquals(401, ex.getHttpResponseException().getStatusCode());
+      Assert.assertEquals(5, authServer.getTotalResponsesServed());
+      Assert.assertEquals(6, registry.getTotalResponsesServed());
+    }
   }
 
   @Test

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryClientTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryClientTest.java
@@ -230,6 +230,8 @@ public class RegistryClientTest {
     } catch (RegistryUnauthorizedException ex) {
       Assert.assertEquals(401, ex.getHttpResponseException().getStatusCode());
       Assert.assertEquals(5, authServer.getTotalResponsesServed());
+      // 1 response asking to do bearer auth + 4 unauth responses for 4 refresh attempts + 1 final
+      // unauth response propagated as RegistryUnauthorizedException here
       Assert.assertEquals(6, registry.getTotalResponsesServed());
     }
   }

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -9,6 +9,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed authentication failure with error `server did not return 'WWW-Authenticate: Bearer' header` in certain cases (for example, on OpenShift). ([#2258](https://github.com/GoogleContainerTools/jib/issues/2258))
+
 ## 2.0.0
 
 ### Added

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -9,6 +9,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed authentication failure with error `server did not return 'WWW-Authenticate: Bearer' header` in certain cases (for example, on OpenShift). ([#2258](https://github.com/GoogleContainerTools/jib/issues/2258))
+
 ## 2.0.0
 
 ### Added

